### PR TITLE
[UNDERTOW-2033] Derive SecurePredicate from connection encryption, r…

### PR DESCRIPTION
**Pull request by baranowb**
…ather than scheme
Issue: https://issues.redhat.com/browse/UNDERTOW-2033
2.3.x:  https://github.com/undertow-io/undertow/pull/1627
2.2.x: https://github.com/undertow-io/undertow/pull/1627

### CODE REVIEW

**Comment by freedge** [comment]
this is how I tested this

Needed this to run the tests:
```
diff --git a/examples/src/main/java/io/undertow/examples/http2/Http2Server.java b/examples/src/main/java/io/undertow/examples/http2/Http2Server.java
index 4ac9b5e..79a2989 100644
--- a/examples/src/main/java/io/undertow/examples/http2/Http2Server.java
+++ b/examples/src/main/java/io/undertow/examples/http2/Http2Server.java
@@ -65,7 +65,7 @@ public class Http2Server {
     public static void main(final String[] args) throws Exception {
         String version = System.getProperty("java.version");
         System.out.println("Java version " + version);
-        if(version.charAt(0) == '1' && Integer.parseInt(version.charAt(2) + "") < 8 ) {
+        if(version.charAt(0) == '1' && version.charAt(1) == '.' && Integer.parseInt(version.charAt(2) + "") < 8 ) {
             System.out.println("This example requires Java 1.8 or later");
             System.out.println("The HTTP2 spec requires certain cyphers that are not present in older JVM's");
             System.out.println("See section 9.2.2 of the HTTP2 specification for details");
diff --git a/pom.xml b/pom.xml
index 36bb250..4ed6370 100644
--- a/pom.xml
+++ b/pom.xml
@@ -111,6 +111,8 @@
         <maven.javadoc.plugin.quiet>true</maven.javadoc.plugin.quiet>
         <maven.compiler.showDeprecation>false</maven.compiler.showDeprecation>
         <maven.compiler.showWarnings>false</maven.compiler.showWarnings>
+
+       <version.compiler.plugin>3.13.0</version.compiler.plugin>
     </properties>

     <modules>
```


I also prepared a new certificate as the one bundled here expired
```
#!/bin/bash

set -e
set -x
rm -f server.keystore client.keystore server.truststore client.truststore server.key server.p12 server.pem
openssl req -x509 -newkey ec -pkeyopt ec_paramgen_curve:secp384r1 -days 3650 \
    -nodes -keyout server.key -out server.pem -subj "/CN=localhost" \
    -addext "subjectAltName=DNS:localhost"
openssl pkcs12 -export -in server.pem -inkey server.key -name server -out server.p12 -passout file:<(echo pass)

keytool -destkeystore server.keystore -deststorepass:file <(echo password) -importkeystore -noprompt -srckeystore server.p12 -srcstorepass:file <(echo pass)
keytool -keystore client.keystore -storepass:file <(echo password) -importcert -noprompt -file server.pem
keytool -keystore server.truststore -storepass:file <(echo password) -importcert -noprompt -file server.pem
keytool -keystore client.truststore -storepass:file <(echo password) -importcert -noprompt -file server.pem
rm -f server.p12 server.key
```

With a recent version of curl you can actually provide the scheme through a header argument, no need for the haproxy headache.

Without the patch:
```
$ build/bin/curl https://localhost:8443 --cacert ~/GIT/undertow/examples/src/main/java/io/undertow/examples/http2/server.pem  -s -o /dev/null -w "%{http_code}\n"
200
$ build/bin/curl https://localhost:8443 --cacert ~/GIT/undertow/examples/src/main/java/io/undertow/examples/http2/server.pem  -s -o /dev/null -w "%{http_code}\n" -H ':scheme: http'
307
$ build/bin/curl http://localhost:8080 --http2-prior-knowledge  -s -o /dev/null -w "%{http_code}\n"
307
$ build/bin/curl http://localhost:8080 --http2-prior-knowledge  -s -o /dev/null -w "%{http_code}\n" -H ':scheme: https'
200
```

Using this patch
```
$ build/bin/curl https://localhost:8443 --cacert ~/GIT/undertow/examples/src/main/java/io/undertow/examples/http2/server.pem  -s -o /dev/null -w "%{http_code}\n"
200
$ build/bin/curl https://localhost:8443 --cacert ~/GIT/undertow/examples/src/main/java/io/undertow/examples/http2/server.pem  -s -o /dev/null -w "%{http_code}\n" -H ':scheme: http'
200
$ build/bin/curl http://localhost:8080 --http2-prior-knowledge  -s -o /dev/null -w "%{http_code}\n"
307
$ build/bin/curl http://localhost:8080 --http2-prior-knowledge  -s -o /dev/null -w "%{http_code}\n" -H ':scheme: https'
307
$
```
It fixes the issue indeed, thanks!

**Review by fl4via — CHANGES_REQUESTED** [review]
ILGTM, just one point can you please fix the typo in the PR message?

File: core/src/main/java/io/undertow/predicate/SecurePredicate.java [hunk 1]
```diff
@@ -33,7 +36,10 @@ public class SecurePredicate implements Predicate {
 
     @Override
     public boolean resolve(HttpServerExchange value) {
-        return value.getRequestScheme().equals("https");
+        final SSLSession session = value.getConnection().getSslSession();
+        final SSLSessionInfo info = value.getConnection().getSslSessionInfo();
+        return (session != null && session.isValid())
+                || (info != null && info.getSSLSession() != null && info.getSSLSession().isValid());
```
Comment (msfm) [inline core/src/main/java/io/undertow/predicate/SecurePredicate.java:42]: Do we need to check `SSLSessionInfo` as well? I thought checking `SSLSession` would be sufficient for this use case. Am I misunderstanding or missing anything?

Comment (baranowb) [reply core/src/main/java/io/undertow/predicate/SecurePredicate.java:42]: Why not both? ( I think in source, though not sure where, I found similar check)

**Comment by baranowb** [comment]
@fl4via done.

**Review by ropalka — APPROVED** [review]

**Pull request merged**